### ISSUE TITLE
BEEFY can get stuck when a majority of validators is restarted

### ISSUE THREAD

**Issue by serban300** [paritytech/polkadot-sdk#3003]
If a majority of the validators are restarted while voting for a mandatory block and before a finality proof was created, it can happen for BEEFY to get stuck at that block since:
- there's a big chance that not all in-flight votes will reach the remaining online validators
- the validators memorize that they voted for block #x, so they won't vote again

Even though this shouldn't happen in production, we should address it since it can affect test nets

**Comment by serban300** [comment]
Another issue is that in this situation, after restart some validators skipped mandatory blocks. We should investigate this too.

**Comment by serban300** [comment]
Managed to reproduce the problem locally with a zombienet. Found 2 issues so far:

1. Finality notifications can be missed. For example:

```
2024-01-23 10:17:59.443 DEBUG tokio-runtime-worker beefy: 🥩 Finality notification: header Header { parent_hash: 0xe7fd79fd15dc7267dc2390b65ea5b64de6e6ce1c484109546d138ec8ee1eb0bc, number: 8, state_root: 0xe15afa338105d137f426feab0bff910e54ab646fa745fe69b320fea43d70ea56, extrinsics_root: 0xdeaf2557f4880fd5ceb4fe53e7ddb5f56f10ee0c690294331a6bc593e921ebe4, digest: Digest { logs: [DigestItem::PreRuntime([66, 65, 66, 69], [3, 5, 0, 0, 0, 9, 150, 242, 16, 0, 0, 0, 0, 20, 90, 70, 34, 25, 118, 177, 171, 56, 148, 250, 60, 180, 143, 44, 205, 78, 7, 220, 242, 189, 241, 223, 7, 101, 254, 233, 67, 130, 120, 229, 0, 186, 45, 91, 22, 206, 91, 109, 139, 36, 13, 114, 10, 166, 218, 54, 64, 5, 224, 239, 232, 227, 113, 121, 200, 235, 149, 203, 114, 107, 7, 191, 7, 196, 148, 187, 3, 200, 98, 73, 52, 50, 131, 39, 6, 6, 248, 187, 134, 170, 57, 4, 155, 254, 29, 237, 12, 32, 23, 76, 236, 224, 99, 55, 2]), DigestItem::Consensus([66, 69, 69, 70], [3, 225, 234, 5, 51, 162, 99, 217, 187, 147, 237, 192, 156, 215, 49, 181, 202, 224, 173, 230, 102, 111, 172, 22, 176, 98, 139, 9, 241, 81, 60, 10, 55]), DigestItem::Seal([66, 65, 66, 69], [168, 140, 233, 174, 61, 155, 31, 250, 167, 34, 230, 246, 114, 233, 201, 108, 143, 205, 244, 29, 203, 49, 0, 236, 180, 98, 185, 103, 152, 154, 174, 88, 99, 169, 227, 251, 186, 23, 8, 253, 25, 226, 216, 15, 25, 206, 116, 215, 152, 75, 228, 109, 137, 216, 31, 239, 77, 34, 239, 209, 6, 139, 221, 142])] } } tree_route []    
2024-01-23 10:17:59.443 DEBUG tokio-runtime-worker beefy: 🥩 New gossip filter: start 1, end 8, validator set id 0    
2024-01-23 10:18:00.026  INFO tokio-runtime-worker babe: 👶 New epoch 1 launching at block 0x7751…a1c8 (block slot 284333580 >= start slot 284333580).    
2024-01-23 10:18:00.026  INFO tokio-runtime-worker babe: 👶 Next epoch starts at slot 284333590    
2024-01-23 10:18:00.031  INFO tokio-runtime-worker substrate: ✨ Imported #11 (0x7751…a1c8)    
2024-01-23 10:18:01.961  INFO tokio-runtime-worker substrate: 💤 Idle (9 peers), best: #11 (0x7751…a1c8), finalized #8 (0x8fa2…210d), ⬇ 6.7kiB/s ⬆ 10.5kiB/s    
2024-01-23 10:18:03.448 DEBUG tokio-runtime-worker beefy: 🥩 Finality notification: header Header { parent_hash: 0x8fa2c02f6ccb4f2fa8599af0889c95921b531d273465d94065387b3f4f9f210d, number: 9, state_root: 0x6977234c8adb2264f4a68d984f3f08838299112da9ebe689e65636166ac07e42, extrinsics_root: 0xa185c93f365d00de093d4ce7764dce252805bb95eb34acb4a50eaf776dd3f5de, digest: Digest { logs: [DigestItem::PreRuntime([66, 65, 66, 69], [3, 5, 0, 0, 0, 10, 150, 242, 16, 0, 0, 0, 0, 108, 202, 30, 17, 135, 51, 126, 231, 154, 173, 194, 143, 86, 244, 43, 80, 214, 173, 198, 183, 225, 158, 178, 58, 85, 204, 132, 121, 121, 93, 248, 70, 124, 23, 19, 94, 153, 253, 85, 138, 77, 20, 52, 162, 223, 57, 179, 201, 134, 182, 90, 221, 221, 188, 243, 127, 240, 35, 76, 200, 136, 203, 82, 1, 98, 72, 0, 227, 250, 106, 143, 136, 252, 67, 219, 80, 122, 51, 116, 139, 84, 87, 206, 149, 72, 54, 158, 94, 218, 33, 134, 152, 191, 178, 234, 0]), DigestItem::Consensus([66, 69, 69, 70], [3, 7, 232, 84, 137, 224, 145, 136, 208, 193, 62, 177, 182, 174, 232, 200, 219, 116, 112, 107, 91, 200, 63, 207, 9, 0, 75, 195, 58, 229, 5, 165, 245]), DigestItem::Seal([66, 65, 66, 69], [110, 121, 203, 219, 208, 160, 0, 219, 85, 206, 45, 5, 130, 107, 138, 16, 118, 201, 21, 217, 41, 141, 188, 84, 78, 253, 128, 27, 144, 19, 8, 3, 126, 165, 85, 234, 210, 177, 79, 9, 139, 180, 148, 244, 246, 196, 194, 166, 36, 99, 142, 248, 55, 12, 34, 197, 9, 82, 176, 187, 5, 207, 237, 128])] } } tree_route []    
2024-01-23 10:18:03.449 DEBUG tokio-runtime-worker beefy: 🥩 New gossip filter: start 1, end 9, validator set id 0    
2024-01-23 10:18:03.449 DEBUG tokio-runtime-worker beefy: 🥩 Try voting on 9    
2024-01-23 10:18:03.449 DEBUG tokio-runtime-worker beefy: 🥩 Local authority id: Public(039ec7147e2859ec2659d8165f50e62bf2300fd62aefafd633e3ed9c8ec9d08344 (1MxAUrwo...))    
2024-01-23 10:18:03.452 DEBUG tokio-runtime-worker beefy: 🥩 Sent vote message: VoteMessage { commitment: Commitment { payload: Payload([([109, 104], [7, 232, 84, 137, 224, 145, 136, 208, 193, 62, 177, 182, 174, 232, 200, 219, 116, 112, 107, 91, 200, 63, 207, 9, 0, 75, 195, 58, 229, 5, 165, 245])]), block_number: 9, validator_set_id: 0 }, id: Public(039ec7147e2859ec2659d8165f50e62bf2300fd62aefafd633e3ed9c8ec9d08344 (1MxAUrwo...)), signature: Signature(4a6a51cb24f033e1fbf671f5f1737395d752d8ea7355bb2a12bf20d33e0739c92c0770f563c0aadb3f375f9d42e55aa9b1890102547ffd4c71730fd80eb899b501) }    
2024-01-23 10:18:03.454  INFO tokio-runtime-worker beefy: 🥩 Round #9 concluded, finality_proof: V1(SignedCommitment { commitment: Commitment { payload: Payload([([109, 104], [7, 232, 84, 137, 224, 145, 136, 208, 193, 62, 177, 182, 174, 232, 200, 219, 116, 112, 107, 91, 200, 63, 207, 9, 0, 75, 195, 58, 229, 5, 165, 245])]), block_number: 9, validator_set_id: 0 }, signatures: [Some(Signature(2e6b4d32a8eb8e4caa3f0f071c598c61acc063edff215bf4aa03928b96093ad7101b4643a1b4c444273488565e1d782b4bb1516fa1d2aa7f2f91749cfb6b13a500)), Some(Signature(43e828391055739e31b829ade281c4fc7b28973f9216da13435ab3c54a87223f75dbd86db29c81954057135f1b92b6bd56f0d13a1199fc3c68eb341907595c1300)), None, Some(Signature(4a6a51cb24f033e1fbf671f5f1737395d752d8ea7355bb2a12bf20d33e0739c92c0770f563c0aadb3f375f9d42e55aa9b1890102547ffd4c71730fd80eb899b501)), Some(Signature(d43a0d4ac389910744f0a3e89bed02ba3b2d4fe7b5076b8abac23b4a03b3bc926aec7a0bfbe1467f5a386cfc7befc99541f8d6395f07a664960af359dbd7c8a301)), None, Some(Signature(3256c71e5f570015b0c52109c0181bdf62b019c2c6a0352ebcdcb157eeb84cd853c9ce96f79765b707fd3dba026510769ddfa3ab8c5b3cae03ee9c83cfbdb9a800)), Some(Signature(b1e977fd842e460f9a14a65fbee095c49911c9d4ff0634e3d28ac23a21bf9be47db8e41f911c48af46f186aae2b4cffcad86b778bae1995f4f09b17a4a32deb600)), Some(Signature(e90bb4fe5d56154898bf328d6d2e2479ba644144c58ac897e72d73ec3b72c0821923406de669aa418c65836cae9e2301c104712c8165209b82f9ad91d6141f5901)), None] }).    
2024-01-23 10:18:03.454 DEBUG tokio-runtime-worker beefy: 🥩 Concluded optional round #9    
2024-01-23 10:18:03.454 DEBUG tokio-runtime-worker beefy: 🥩 New gossip filter: start 9, end 9, validator set id 0    
2024-01-23 10:18:03.454 DEBUG tokio-runtime-worker beefy: 🥩 received vote for old stale round 9, ignoring    
2024-01-23 10:18:03.457 DEBUG tokio-runtime-worker beefy: 🥩 Process justification for round: 9.    
2024-01-23 10:18:03.457 DEBUG tokio-runtime-worker beefy: 🥩 received vote for old stale round 9, ignoring    
2024-01-23 10:18:03.457 DEBUG tokio-runtime-worker beefy: 🥩 received vote for old stale round 9, ignoring    
tail: /home/serban/workplace/snapshots/db-test-gen-test/validator-unstable-0.log: file truncated
2024-01-23 10:18:08.644  INFO main sc_cli::runner: Parity Polkadot    
2024-01-23 10:18:08.644  INFO main sc_cli::runner: ✌️  version 1.6.0-c83ccda4a32    
2024-01-23 10:18:08.644  INFO main sc_cli::runner: ❤️  by Parity Technologies <admin@parity.io>, 2017-2024    
2024-01-23 10:18:08.644  INFO main sc_cli::runner: 📋 Chain specification: Rococo Local Testnet    
2024-01-23 10:18:08.644  INFO main sc_cli::runner: 🏷  Node name: validator-unstable-0    
2024-01-23 10:18:08.644  INFO main sc_cli::runner: 👤 Role: AUTHORITY    
2024-01-23 10:18:08.644  INFO main sc_cli::runner: 💾 Database: RocksDb at /home/serban/workplace/snapshots/db-test-gen-test//validator-unstable-0/data/chains/rococo_local_testnet/db/full    
2024-01-23 10:18:10.516 DEBUG main beefy: 🥩 Registered block-import metrics    
2024-01-23 10:18:16.142 DEBUG main beefy: 🥩 Registered on-demand incoming justification requests metrics    
2024-01-23 10:18:16.160  INFO main sub-libp2p: 🏷  Local node identity is: 12D3KooWCcdeufa44tdNLPxQVXMSmv6grzd4vp8ZDP7RjrXViqws    
2024-01-23 10:18:16.392  INFO main polkadot_service: 🚀 Using prepare-worker binary at: "/media/serban/data/workplace/sources/polkadot-sdk/target/release/polkadot-prepare-worker"    
2024-01-23 10:18:16.392  INFO main polkadot_service: 🚀 Using execute-worker binary at: "/media/serban/data/workplace/sources/polkadot-sdk/target/release/polkadot-execute-worker"    
2024-01-23 10:18:16.421  INFO main sc_sysinfo: 💻 Operating system: linux    
2024-01-23 10:18:16.421  INFO main sc_sysinfo: 💻 CPU architecture: x86_64    
2024-01-23 10:18:16.421  INFO main sc_sysinfo: 💻 Target environment: gnu    
2024-01-23 10:18:16.421  INFO main sc_sysinfo: 💻 CPU: 12th Gen Intel(R) Core(TM) i7-12700H    
2024-01-23 10:18:16.421  INFO main sc_sysinfo: 💻 CPU cores: 14    
2024-01-23 10:18:16.421  INFO main sc_sysinfo: 💻 Memory: 39793MB    
2024-01-23 10:18:16.421  INFO main sc_sysinfo: 💻 Kernel: 6.5.0-14-generic    
2024-01-23 10:18:16.421  INFO main sc_sysinfo: 💻 Linux distribution: Ubuntu 23.10    
2024-01-23 10:18:16.421  INFO main sc_sysinfo: 💻 Virtual machine: no    
2024-01-23 10:18:16.421  INFO main sc_service::builder: 📦 Highest known block at #11    
2024-01-23 10:18:16.422  INFO tokio-runtime-worker substrate_prometheus_endpoint: 〽️ Prometheus exporter started at 0.0.0.0:37481    
2024-01-23 10:18:16.423  INFO                 main sc_rpc_server: Running JSON-RPC server: addr=0.0.0.0:41029, allowed origins=["*"]    
2024-01-23 10:18:16.423  INFO                 main sc_sysinfo: 🏁 CPU score: 732.66 MiBs    
2024-01-23 10:18:16.423  INFO                 main sc_sysinfo: 🏁 Memory score: 2.62 GiBs    
2024-01-23 10:18:16.423  INFO                 main sc_sysinfo: 🏁 Disk score (seq. writes): 662.47 MiBs    
2024-01-23 10:18:16.423  INFO                 main sc_sysinfo: 🏁 Disk score (rand. writes): 272.66 MiBs    
2024-01-23 10:18:16.423  WARN                 main polkadot_service: ⚠️  The hardware does not meet the minimal requirements Failed checks: BLAKE2-256(expected: 783.27 MiBs, found: 732.66 MiBs), Copy(expected: 11.49 GiBs, found: 2.62 GiBs), Seq Write(expected: 950.00 MiBs, found: 662.47 MiBs), Rnd Write(expected: 420.00 MiBs, found: 272.66 MiBs),  for role 'Authority' find out more at:
https://wiki.polkadot.network/docs/maintain-guides-how-to-validate-polkadot#reference-hardware    
2024-01-23 10:18:16.436  INFO                 main babe: 👶 Starting BABE Authorship worker    
2024-01-23 10:18:16.439 DEBUG tokio-runtime-worker beefy: 🥩 Registered voter metrics    
2024-01-23 10:18:16.439 DEBUG tokio-runtime-worker beefy: 🥩 Registered on-demand outgoing justification requests metrics    
2024-01-23 10:18:16.439  INFO tokio-runtime-worker beefy: 🥩 BEEFY gadget waiting for BEEFY pallet to become available...    
2024-01-23 10:18:16.485  INFO tokio-runtime-worker substrate: ✨ Imported #12 (0x4e1a…43f4)    
2024-01-23 10:18:16.505  INFO tokio-runtime-worker substrate: ✨ Imported #13 (0x9307…27b0)    
2024-01-23 10:18:18.025  INFO tokio-runtime-worker substrate: ✨ Imported #14 (0xfc57…1d63)    
2024-01-23 10:18:20.595  INFO tokio-runtime-worker beefy: 🥩 BEEFY pallet available: block 12 beefy genesis 1    
2024-01-23 10:18:20.595  INFO tokio-runtime-worker beefy: 🥩 Loading BEEFY voter state from db: PersistedState { best_voted: 9, voting_oracle: VoterOracle { sessions: [Rounds { rounds: {}, previous_votes: {}, session_start: 1, validator_set: ValidatorSet { validators: [Public(03920968eb41e5acfa8dc9ff3a994d91e4e07fd7172785b6f8566a7be07a34fc90 (1MfTZbKZ...)), Public(03c484cf54465a8ab90f64a29052008843fa627ef7e39aa55d3e8f63bd685e41b3 (1NoecbFd...)), Public(03ee9147bf713be07cc23a2416a939adc8f9d0bf68a2746a07c2d81dc4d991e605 (1PknL6em...)), Public(039ec7147e2859ec2659d8165f50e62bf2300fd62aefafd633e3ed9c8ec9d08344 (1MxAUrwo...)), Public(020fda8804f49949747ed4dbe2a40f97b427446da079cd87ee8a1465727b664017 (1Cw79rMe...)), Public(03f0307642a73cbf89ce10e95b4aa37175471faa57b6555cfd29e27649fafe0869 (1PnufWfN...)), Public(02bb5b3068b9290ec3be5bac40ac2a286f9ecd5276083d2b48f76b82a12f42f4cb (1GoyZVcN...)), Public(0324633b13e945726747be10621b837358ef7033774db6848520094c3353f06f79 (1KBgyfdG...)), Public(03ea6c68d7815c4141ae30ea91a15bc234af654bd7c8034dca9a384f2755032db7 (1PfMBinG...)), Public(0399a14902a0461dd8a4eaa002d225c1a3e4fffa3354f8136d144b9750f10d46f7 (1MqR1nVg...))], id: 0 }, mandatory_done: true, best_done: Some(9) }], min_block_delta: 8, best_grandpa_block_header: Header { parent_hash: 0x77510e432606ac27cdb2989787ab6c5ecdb5065a30e8e110fef133ee575aa1c8, number: 12, state_root: 0xf02fb24c6ba2123563578e674a11725d924da3725a54d724228d29c9566ce3a9, extrinsics_root: 0x2a0be8e5a6b5410ed0822c3f34c271936b398b2b8c451deffe454b8953004a6a, digest: Digest { logs: [DigestItem::PreRuntime([66, 65, 66, 69], [3, 1, 0, 0, 0, 13, 150, 242, 16, 0, 0, 0, 0, 206, 73, 35, 217, 78, 128, 152, 202, 144, 214, 152, 210, 25, 1, 251, 243, 248, 16, 73, 8, 43, 56, 89, 186, 65, 50, 92, 25, 29, 208, 74, 63, 213, 102, 225, 74, 175, 223, 98, 96, 39, 4, 17, 141, 173, 232, 42, 109, 228, 239, 33, 153, 79, 150, 36, 44, 55, 73, 197, 78, 167, 175, 95, 3, 119, 222, 255, 87, 168, 137, 115, 93, 226, 230, 57, 65, 246, 185, 92, 71, 149, 159, 178, 59, 232, 32, 53, 217, 99, 150, 124, 238, 241, 219, 126, 10]), DigestItem::Consensus([66, 69, 69, 70], [3, 176, 6, 170, 91, 89, 101, 223, 211, 183, 81, 196, 16, 75, 125, 84, 102, 169, 187, 43, 28, 10, 245, 107, 59, 132, 48, 50, 208, 201, 156, 84, 129]), DigestItem::Seal([66, 65, 66, 69], [188, 218, 126, 25, 215, 252, 92, 150, 104, 254, 185, 25, 104, 25, 210, 54, 152, 41, 144, 243, 180, 194, 27, 238, 250, 45, 16, 8, 213, 52, 229, 99, 229, 23, 255, 9, 73, 141, 193, 11, 109, 171, 32, 233, 7, 65, 29, 11, 145, 77, 8, 43, 151, 135, 180, 115, 177, 208, 38, 166, 71, 110, 30, 141])] } }, best_beefy_block: 9 }, pallet_genesis: 1 }.    
2024-01-23 10:18:20.595 DEBUG tokio-runtime-worker beefy: 🥩 New gossip filter: start 9, end 12, validator set id 0    
2024-01-23 10:18:20.595  INFO tokio-runtime-worker mmr: Loading MMR best canonicalized state from db: 9.    
2024-01-23 10:18:20.595  INFO tokio-runtime-worker beefy: 🥩 run BEEFY worker, best grandpa: #12.    
2024-01-23 10:18:21.424  INFO tokio-runtime-worker substrate: 💤 Idle (9 peers), best: #14 (0xfc57…1d63), finalized #12 (0x4e1a…43f4), ⬇ 33.4kiB/s ⬆ 25.2kiB/s    
2024-01-23 10:18:24.026  INFO tokio-runtime-worker substrate: ✨ Imported #15 (0xa8ab…0dd8)    
2024-01-23 10:18:26.425  INFO tokio-runtime-worker substrate: 💤 Idle (9 peers), best: #15 (0xa8ab…0dd8), finalized #12 (0x4e1a…43f4), ⬇ 7.8kiB/s ⬆ 9.4kiB/s    
2024-01-23 10:18:28.603 DEBUG tokio-runtime-worker beefy: 🥩 Finality notification: header Header { parent_hash: 0x4e1a6adec1b3db31650076fa427347c7470a7426627dd8b84b3cc98efa2843f4, number: 13, state_root: 0xd165c1366d2482fa33ebaba82e00d0eb0040fd72d4ded80971ff4d646a3afe8a, extrinsics_root: 0x1fcd4a19de94583cf383749f5b2b4b96555b9492dc8d118bc52fbedabf597f26, digest: Digest { logs: [DigestItem::PreRuntime([66, 65, 66, 69], [3, 0, 0, 0, 0, 14, 150, 242, 16, 0, 0, 0, 0, 134, 219, 157, 251, 128, 139, 9, 62, 69, 29, 1, 192, 127, 143, 80, 255, 148, 27, 164, 94, 194, 82, 87, 11, 139, 28, 226, 178, 21, 112, 187, 79, 185, 216, 253, 154, 196, 178, 185, 125, 85, 208, 34, 200, 3, 196, 186, 214, 217, 39, 233, 14, 180, 231, 93, 55, 173, 111, 217, 137, 101, 169, 182, 5, 243, 129, 2, 229, 20, 92, 58, 138, 42, 61, 153, 92, 14, 34, 185, 158, 181, 82, 97, 32, 53, 189, 116, 119, 246, 176, 12, 12, 225, 42, 27, 4]), DigestItem::Consensus([66, 69, 69, 70], [3, 91, 71, 41, 217, 53, 47, 144, 66, 224, 101, 247, 153, 5, 179, 166, 224, 167, 130, 228, 56, 187, 61, 49, 57, 167, 4, 41, 2, 115, 121, 130, 81]), DigestItem::Seal([66, 65, 66, 69], [60, 220, 39, 180, 133, 76, 201, 231, 182, 11, 252, 77, 175, 48, 208, 170, 97, 30, 53, 244, 178, 58, 87, 162, 77, 141, 84, 230, 60, 171, 227, 16, 9, 142, 63, 150, 115, 44, 154, 92, 135, 26, 15, 100, 156, 136, 158, 178, 42, 222, 133, 254, 111, 192, 63, 203, 198, 191, 232, 50, 180, 43, 144, 139])] } } tree_route []    
2024-01-23 10:18:28.604 DEBUG tokio-runtime-worker beefy: 🥩 New gossip filter: start 9, end 13, validator set id 0    
```

Here the finality notification for block 11 is missed. I don't know if before the restart or immediately after, but BEEFY doesn't get it and continues with block 12, not knowing that 11 is mandatory.

2. [`wait_for_runtime()`](https://github.com/paritytech/polkadot-sdk/blob/8eeacff484cf7c7ec6081495dcaffef5a0867bc5/substrate/client/consensus/beefy/src/lib.rs#L298) waits for the next finality notification (let's say #x) and checks if that block has the beefy pallet. But after that if we also have a voter state in auxdb, we don't add #x to the voting rounds, even if #x is mandatory

So probably we should do the BEEFY walk back, even if we have a persisted state

**Comment by serban300** [comment]
Another issue is that `voting_target()` can return blocks from future sessions. So even if the current session is not finalized, because the mandatory block is not finalized for example, the worker still goes on and votes for blocks from following sessions

**Comment by serban300** [comment]
Merged #3074 into master. Having this, in my local tests the problem doesn't reproduce anymore. I don't know if we need to also rebroadcast votes. Thinking.
